### PR TITLE
[ntuple] Initial implementation of bulk I/O

### PR DESCRIPTION
**Pull request by jblomer**

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 1]
```diff
@@ -174,6 +174,83 @@ public:
...
+         auto nRead = fField->ReadBulk(bulkSpec);
+         if (nRead == RBulkSpec::kAllSet) {
+            if ((offset == 0) && (size == fSize)) {
+               fNValidValues = fSize;
+            } else {
+               CountValidValues();
+            }
+         } else {
+            fNValidValues += nRead;
+         }
```
Comment (eguiraud) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:249]: if `ReadBulk` returns the number of values for which the mask was flipped from false to true, could these lines just be 
```suggestion
         fNValidValues += nRead;
```
instead?

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 2]
```diff
@@ -312,6 +404,26 @@ protected:
...
+
+   /// Returns the number of newly available values, that is the number of bools in bulkSpec.fMaskAvail that
+   /// flipped from false to true. As a special return value, kAllSet can be used if all values are read
+   /// independent from the masks.
+   std::size_t ReadBulk(const RBulkSpec &bulkSpec)
+   {
+      if (fIsSimple) {
+         /// For simple types, ignore the mask and memcopy the values into the destination
+         fPrincipalColumn->ReadV(bulkSpec.fFirstIndex, bulkSpec.fCount, bulkSpec.fValues);
+         memset(bulkSpec.fMaskAvail, 1, bulkSpec.fCount);
```
Comment (eguiraud) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:420]: nitpick (I almost didn't add this comment to the review but hey, I typed it, it's now yours to do whatever you want with it): this relies on the fact that `sizeof(bool)` is and will always be 1. I guess if that ever becomes false (as it's not actually guaranteed by the standard) there is _a lot more code_ that will have to be changed, but if you want to be a but more future-proof and declarative:

```suggestion
         std::fill(bulkSpec.fMaskAvail, bulkSpec.fMaskAvail + bulkSpec.fCount, true);
```

It should compile down to a `memset` anyway.

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 3]
```diff
@@ -312,6 +404,26 @@ protected:
...
+         return RBulkSpec::kAllSet;
```
Comment (eguiraud) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:421]: what's the advantage of returning the special value here instead of `bulkSpec.fCount` (which would make for a more consistent API and remove the need for callers to check for the special value)?

File: tree/ntuple/v7/src/RField.cxx [hunk 4]
```diff
@@ -447,6 +523,26 @@ void ROOT::Experimental::Detail::RFieldBase::ReadGlobalImpl(ROOT::Experimental::
...
+   for (std::size_t i = 0; i < bulkSpec.fCount; ++i) {
+      // Value not needed
+      if (!bulkSpec.fMaskReq[i])
+         continue;
+
+      // Value already present
+      if (bulkSpec.fMaskAvail[i])
+         continue;
+
+      Read(bulkSpec.fFirstIndex + i, reinterpret_cast<unsigned char *>(bulkSpec.fValues) + i * valueSize);
```
Comment (eguiraud) [inline tree/ntuple/v7/src/RField.cxx:539]: RBulkSpec stores `fValues` as `void *` while `RBulk` uses `unsigned char *`. I don't have an opinion either way but it would be less surprising to be consistent.

File: tree/ntuple/v7/test/CMakeLists.txt [hunk 5]
```diff
@@ -24,6 +24,7 @@ if(MSVC)
 endif()
 
 ROOT_ADD_GTEST(ntuple_basics ntuple_basics.cxx LIBRARIES ROOTDataFrame ROOTNTuple MathCore CustomStruct)
+ROOT_ADD_GTEST(ntuple_bulk ntuple_bulk.cxx LIBRARIES ROOTDataFrame ROOTNTuple)
```
Comment (eguiraud) [inline tree/ntuple/v7/test/CMakeLists.txt:27]:
```suggestion
ROOT_ADD_GTEST(ntuple_bulk ntuple_bulk.cxx LIBRARIES ROOTNTuple)
```

File: tree/ntuple/v7/test/ntuple_bulk.cxx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      auto fldInt = model->MakeField<int>("int");
+      auto writer = RNTupleWriter::Recreate(std::move(model), "ntpl", fileGuard.GetPath());
+      for (int i = 0; i < 20; ++i) {
+         *fldInt = i;
+         writer->Fill();
+      }
+   }
+
+   auto reader = RNTupleReader::Open("ntpl", fileGuard.GetPath());
+   // TODO(jblomer): find a better way to expose the GenerateBulk method of the target field
```
Comment (eguiraud) [inline tree/ntuple/v7/test/ntuple_bulk.cxx:17]: yeah I would never have guessed this was the way to do it now :sweat_smile:

File: tree/ntuple/v7/test/ntuple_bulk.cxx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      bulk = std::make_unique<RFieldBase::RBulk>(f.GenerateBulk());
+   }
+
+   auto mask = std::make_unique<bool[]>(10);
+   memset(mask.get(), 1, 10);
+   auto intArr = static_cast<int *>(bulk->ReadBulk(RClusterIndex(0, 0), mask.get(), 10));
+   for (int i = 0; i < 10; ++i) {
+      EXPECT_EQ(i, intArr[i]);
+   }
+}
```
Comment (eguiraud) [inline tree/ntuple/v7/test/ntuple_bulk.cxx:32]: it would be good to add a test that reads a collection. it's good to test that the resulting array elements are all contained in a flattened slab of memory, which is something RDF might rely on in the future.

**Review by eguiraud — APPROVED** [review]
Nice, LGTM!

The code in general is easy to follow and unsurprising (which is great!). However it's a bit hard to follow which class is a nested class of which other class and what methods (or  nested classes!) are accessible to what other classes via `protected` access specifications and friendships. We might not need that level of fine-grained encapsulation in namespace `Detail`.

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 8]
```diff
@@ -174,6 +174,83 @@ public:
...
+      ~RBulk();
+      RBulk(const RBulk &) = delete;
+      RBulk(RBulk &&other);
+      RBulk &operator=(RBulk &&other);
+
+      /// Reads 'size' values from the associated field, starting from 'firstIndex'. Note that the index is given
+      /// relative to a certain cluster. The return value points to the array of read objects.
+      /// The 'maskReq' parameter is a bool array of at least 'size' elements. Only objects for which the mask is
+      /// true are guaranteed to be read in the returned value array.
+      void *ReadBulk(const RClusterIndex &firstIndex, const bool *maskReq, std::size_t size)
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:227]: My idea was to inline the simple case, which is:

1. The range of values requested is already available, or
2. Bulk I/O on a simple type

File: tree/ntuple/v7/inc/ROOT/RField.hxx (hunk 1 above)
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:253]: Connected to the question below: At least in principle it is possible that some of the values are already available. If we simply return `bulkSpec.fCount`, we would double count these bits.  On the other hand, I would like to avoid counting the bits in the mask for the common cases. Hence the more complicated logic (which has indeed some smell to it).

File: tree/ntuple/v7/test/CMakeLists.txt (hunk 5 above)
Comment (jblomer) [reply tree/ntuple/v7/test/CMakeLists.txt:27]: In principle, yes, but the utility header `ntuple_test.hxx` depends on it. Perhaps this header can be split in another PR.

File: tree/ntuple/v7/test/ntuple_bulk.cxx (hunk 7 above)
Comment (jblomer) [reply tree/ntuple/v7/test/ntuple_bulk.cxx:91]: Yep, that can be done after adding the `ReadBulk` version for `RRVecField`. I wanted to do this in a follow-up PR but I can also do it here.

File: tree/ntuple/v7/inc/ROOT/RField.hxx (hunk 1 above)
Comment (eguiraud) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:253]: I think I get what you mean but it does not seem possible in practice that kAll is returned and some of the values were already available: kAll is used by implementations of bulk reading that always read all values.

File: tree/ntuple/v7/inc/ROOT/RField.hxx (hunk 8 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:227]: makes total sense, thanks!
